Mk transifex pull

mkangia opened the pull request:
Second phase for transifex integration after https://github.com/dimagi/commcare-hq/pull/20492
This PR:
1. provides the view changes to pull translated files from transifex
2. checks if all has been translated before pulling files. Provided an option for one to skip that test if needed for now
3. Generates an excel file just like one would download for app translations and emails it to the user

Probably better to review all at once.

mkangia committed:
add ability to pull translations from transifex

mkangia committed:
remove redundant message

mkangia committed:
used temporary file instead

mkangia committed:
lint

mkangia committed:
add ability to check translation stats for a resource

mkangia committed:
add check for translated resources on view when pulling

mkangia committed:
lint [skip ci]

mkangia committed:
improvements and documentation

esoergel commented on custom/icds/translations/integrations/transifex.py:
```diff
@@ -83,3 +93,64 @@ def _cleanup(self):
             if os.path.exists(filepath):
                 os.remove(filepath)
+
+    def _get_resource_slugs_for_version(self, version):
```
It looks like `version` is always just `self.version`.  Would it be possible to make a memoized property that's just `resource_slugs`?  Then you wouldn't need to pass it around as much.

mkangia replied:
Ya, removed passing version argument.
> memoized property that's just resource_slugs

sorry, could not get this.

esoergel replied:
haha, it took me a minute to decipher my own words 3 days later here.  I meant that you could replace
```python
def _get_resource_slugs_for_version(self, version):
```
with
```python
@property
@memoized
def resource_slugs(self):
```

esoergel commented on custom/icds/translations/integrations/transifex.py:
```diff
@@ -83,3 +93,64 @@ def _cleanup(self):
+        :return: list of POEntry objects
+        """
+        if resource_slugs:
+            self._ensure_resource_slugs_for_version(resource_slugs, self.version)
```
I think this check should go in the code that fetches the `resource_slugs` - one place that's responsible for returning valid slugs, so everywhere else knows what they're working with.

mkangia replied:
It is called with resource_slugs as `None`. So if there are no resource slugs it tries to find them first and if its still missing, then it raises an `Exception`.

mkangia replied:
`resource_slugs` can be passed to this function as an argument. Currently there is no place where its fetched and then passed to this method.

esoergel commented on corehq/apps/app_manager/app_translations/parser.py:
```diff
@@ -0,0 +1,112 @@
+from corehq.apps.dump_reload.const import DATETIME_FORMAT
+
+
+class TranslationsParser:
```
I think this should be
```python
class TranslationsParser(object):
```

mkangia replied:
hmm, i have started doing this recently. Is there an advantage by using object here?

esoergel commented on custom/icds/translations/integrations/transifex.py:
```diff
@@ -26,14 +28,21 @@ def __init__(self, domain, app_id, source_lang, project_slug, version=None, lang
-        self.po_file_generator = POFileGenerator(domain, app_id, version, key_lang, source_lang, lang_prefix,
+        self.po_file_generator = POFileGenerator(domain, app_id, version, self.key_lang, source_lang, lang_prefix,
                                                  exclude_if_default)
+        self.parser = TranslationsParser(self)
```
This dependency is a little weird - `Transifex` has an attribute called `parser`, which itself has `transifex` as an attribute.
It looks like `transifex.parser` is only called once, for `generate_excel_file`.  An alternative approach would be to make a `generate_excel_file` method on the `Transifex` class like this:
```python
def generate_excel_file(self):
    parser = TranslationsParser(self)
    return parser.generate_excel_file()
```
That way you could change the parser around without needing to worry about the calling code, since it doesn't access it directly.  Further, my preference would be to pass only the dependencies that `TranslationsParser` needs, rather than giving it this whole class.  That would avoid such tight coupling between the two classes.

mkangia replied:
that makes sense.. that circular reference was not a good plan. Updating.

esoergel commented on custom/icds/translations/integrations/transifex.py:
```diff
@@ -83,3 +93,64 @@ def _cleanup(self):
+        """
+        pull translations from transifex
+        :param resource_slugs: optional argument. All resource slugs corresponding to version are used otherwise.
+        :return: list of POEntry objects
```
nit: it returns a dict, not a list

esoergel commented on corehq/apps/app_manager/app_translations/parser.py:
```diff
@@ -0,0 +1,112 @@
+        version = self.transifex.version
+        all_translations = self.transifex.get_translations(version, resource_slugs)
+        for resource_name in all_translations:
+            po_entries = all_translations[resource_name]
```
You could also do:
```python
for resource_name, po_entries in all_translations.items():
```
or just make `all_translations` a list of tuples if you don't need the dictionary functionality anyways.

mkangia replied:
I remember using it and it was causing an issue. let me try again.

list of tuples? How would that be?
need the dict where resource_name is mapped to a list of translations/PoEntries

mkangia replied:
updated to use the for loop with items.

esoergel replied:
> list of tuples? How would that be?

I mean that these two are equivalent:

```python
po_entries = {}
for resource_slug in resource_slugs:
    po_entries[resource_slug] = self.client.get_translation(resource_slug, self.source_lang, self.lock_translations)
return po_entries

for resource_name, po_entries in get_translations().items():
```

```python
po_entries = [] 
for resource_slug in resource_slugs:
    po_entries.append((resource_slug, self.client.get_translation(resource_slug, self.source_lang, self.lock_translations)))
return po_entries

for resource_name, po_entries in get_translations():
```
Using a dictionary lets you jump to specific entries in the set, or more easily modify them after the fact, but in this case, it iterates over the whole set using `items()`, which converts the dictionary into a list of tuples, so you could also just build a list of tuples directly.  They're really pretty much the same though - I don't think it matters.

esoergel reviewed (commented):
Looks like you're passing around tempfiles in a couple places here.  As we recently ran into with exports, you'll need to manually clean these up, including instances where there's an exception thrown somewhere in the middle.  The default way to do this is using `NamedTemporaryFile` in a `with` statement (but without `delete=False`).  This with statement should then span the whole lifecycle of that file.  Once you're done writing, you can then `seek` to 0 on the file and read from the same object.

I also wrote an alternative utility called [`TransientTempfile`](https://github.com/dimagi/commcare-hq/blob/master/corehq/util/files.py#L47-L47) which uses a `with` statement only for the lifecycle of the file, and you handle reads and writes with normal `with open` statements.

There are also a few optional parameters in here that I don't think are necessary.  If you make those mandatory, then you can eliminate a bunch of branching logic that deals with whether or not that was provided.  Essentially, it's a way to make your code stricter and a bit simpler.

mkangia commented on corehq/apps/app_manager/app_translations/parser.py:
```diff
@@ -0,0 +1,112 @@
+            ws = wb.create_sheet(title=sheet_name)
+            self._parse_sheet(ws, po_entries)
+
+    def generate_excel_file(self, resource_slugs=None):
```
Ya, its more of a proactive addition because it would probably come in handy to pull specific resource slugs.  I used it during testing so i left it there for now. Happy to remove if you think its not necessary/YAGNI.

mkangia commented on corehq/apps/app_manager/app_translations/parser.py:
```diff
@@ -0,0 +1,112 @@
+        wb = Workbook(write_only=True)
+        self._generate_sheets(wb, resource_slugs)
+
+        with NamedTemporaryFile(delete=False) as tempfile:
```
Yup, done later.

mkangia commented on custom/icds/translations/integrations/client.py:
```diff
@@ -66,3 +69,31 @@ def project_details(self):
+        response = requests.get(url, auth=self._auth, stream=True)
+        if response.status_code != 200:
+            return response
+        temp_file = tempfile.NamedTemporaryFile()
```
Just to confirm
```
temp_file = tempfile.NamedTemporaryFile()
temp_file.write(response.content)
temp_file.seek(0)
polib_obj = polib.pofile(temp_file.name)
temp_file.close()
```
This looks good?

esoergel replied:
Actually I don't think you'll need the `seek` line.  Since you're not passing the open file reference to `polib`, it must open the file again.  Otherwise that looks correct.

You'll need to make sure `tempfile.close()` doesn't get closed until `polib_obj` is no longer needed, since it'll delete the file.  Similarly, it's safer to put all operations on this temfile inside a `with` statement so the file gets cleaned up even on failure.  That can look like this:
```python
with tempfile.NamedTemporaryFile() as tempfile:
    temp_file.write(response.content)
    polib_obj = polib.pofile(temp_file.name)
```
The downside of this is that `polib_obj` maintains a separate handler to the same file, so you'll have the file open twice.  I believe having a file open multiple times works on most systems (just not Windows, I think?).  It could get confusing if both tried writing to it in sequence.  That `TransientTempfile` util I mentioned in my review comment lets you close the file without yet deleting it.  This use-case is probably simple enough that it won't run into issues either way, though.

mkangia replied:
> Actually I don't think you'll need the seek line.

I tried this and just using `write` didnt write to file, had to do `seek` to make it actually write the content.

> You'll need to make sure tempfile.close() doesn't get closed until polib_obj is no longer needed, since it'll delete the file

Ya i think we don't need that file after that. `polib.pofile(temp_file.name)` parses the file and converts it to `POEntries`. i have updated this now and there is no `po_obj`.

esoergel replied:
> I tried this and just using write didnt write to file, had to do seek to make it actually write the content.

`seek(0)` moves the pointer back to the beginning of the file.  `write` does write to the file, but you can't read from the same file object again without resetting the pointer.  `polib.pofile` only gets the file name, so it must be opening the file a second time, which means it *should* be able to read without seeking.  Here's an example:
```python
In [1]: import tempfile

In [2]: with tempfile.NamedTemporaryFile() as f:
   ...:     f.write('hello')
   ...:     print "read 1", f.read()
   ...:     f.seek(0)
   ...:     print "read 2", f.read()
   ...:     print "read 3", f.read()
   ...:     with open(f.name) as f2:
   ...:         print "read 4", f2.read()
   ...:         
read 1 
read 2 hello
read 3 
read 4 hello
```

> Ya i think we don't need that file after that. polib.pofile(temp_file.name) parses the file and converts it to POEntries. i have updated this now and there is no po_obj.

Cool yeah, as long as that doesn't need the tempfile after initialization this should be fine.  I wasn't sure whether it streamed the file lazily or loaded it all into memory up front, for example.

mkangia replied:
@esoergel tried making this change and ran into some issue with encoding because of it.
Want to leave it as-is for now and look at this improvement later.

Also looks like with the current implementation,
it should be safe to leave as-is. When the `temp_file` reference is removed, the file will get removed as well, so there should be no orphan file references, temp files even if anything goes wrong. let me know if thats okay.

mkangia commented on custom/icds/translations/integrations/transifex.py:
```diff
@@ -83,3 +93,64 @@ def _cleanup(self):
+        if resource_slugs:
+            self._ensure_resource_slugs_for_version(resource_slugs, self.version)
+        client = self.client()
+        if not resource_slugs:
```
It is called with resource_slugs as `None`. So if there are no resource slugs it tries to find them first and if its still missing, then it raises an Exception.

mkangia committed:
review improvments

- make client method a property instead
- avoid adding parser as a property on transifx object and then add transifex object as property on parser
- avoid passing version as argument

mkangia committed:
fix order

mkangia commented:
@esoergel 
Made few changes for the reviews and replied to rest.
Looks like there is some possibility of improving the structure for this. Will try to make some improvements.

mkangia committed:
rename parse -> add

mkangia committed:
improve lang mapping use and make client less visible

Any hq lang code to transifex lang code moved to a single place in client
Make all client actions available through transifex object itself

mkangia committed:
remove redundant translations dict used earlier

mkangia committed:
add comments and minor improvement

mkangia committed:
break it down

mkangia committed:
fix issue with memoizing

memoize failed trying to hash a dict for key for memoizing

mkangia committed:
check for source lang only

mkangia committed:
minor improvements and documentation

mkangia committed:
memoize file name and include generate excel within try

mkangia committed:
lock resources when pulling translations to freeze them

mkangia committed:
add object as parent

https://docs.python.org/2/reference/datamodel.html#new-style-and-classic-classes

mkangia committed:
add ability to delete resources

esoergel commented on custom/icds_reports/views.py:
```diff
@@ -1624,13 +1624,13 @@ def page_context(self):
+            source_language_code = form_data.get('target_lang') or form_data.get('source_lang')
+            self._transifex = Transifex(domain, form_data['app_id'], source_language_code, transifex_project_slug,
+                                        form_data['version'])
+        return self._transifex
```
Looks like this is only accessed directly once, so you could just rename it `get_transifex` and not bother with memoizing the result.

esoergel commented on custom/icds/translations/integrations/transifex.py:
```diff
@@ -83,3 +88,68 @@ def _cleanup(self):
+                                                                    self.lock_translations)
+        return po_entries
+
+    def resources_pending_translations(self, break_if_true=False):
```
This method returns a single slug if `break_if_true`, and a list of slugs if `break_if_false`.  It looks like this is used in only one place, and it's being used as a boolean.  It feels to me like it's one method trying to act like several depending on how it's called.  Considering this is only used in a boolean context, and with `break_if_true=True`, I'd change the name and implementation to reflect that.  That's my preference, but I'll defer to you on the final call.

mkangia replied:
> it's being used as a boolean

the result is used to list the pending resources/resources as well, so `resources_pending_translations` sounds fine. Can make it always return a list, so in case `break_if_true` is `True` return a list with the single resource to make the return predictable.

mkangia committed:
add resource_slugs on transifex as a memoized property and other minor improvements

mkangia committed:
avoid memoizing when not needed

mkangia committed:
lint

esoergel commented on custom/icds/translations/integrations/transifex.py:
```diff
@@ -89,42 +91,33 @@ def _cleanup(self):
+        if self._resource_slugs:
+            return self._resource_slugs
+        else:
+            return self.client.get_resource_slugs(self.version)
```
:+1:
